### DISCUSSION TITLE
Deprecate support for wrapping sync things as async

### DISCUSSION THREAD

**Discussion by adriangb** [encode/starlette#1628] [Q&A]
There are several issues related to wrapping sync things to make them async:
- #1444 
- #1258 
- https://github.com/encode/starlette/discussions/1573#discussion-3987932

Are a couple I can think of.

A comment from Tom: https://github.com/encode/starlette/issues/886#issuecomment-625167713

In the interest of keeping Starlette minimal, I think we should deprecate and later remove this functionality.
Downstream consumers can choose to vendor this functionality (it's only a couple LOC) since they'll have more control over the rough edges by virtue of being less general of a framework.

**Comment by adriangb** [comment]
cc @Kludex

**Reply by Kludex** [reply]
I'm not into this idea. 😕

**Reply by adriangb** [reply]
Yeah that's fair, I don't expected to be without opposition. Would you mind summarizing your thoughts on why we should provide this functionality to users?

**Reply by Kludex** [reply]
I don't think the arguments for removing it are strong enough. There's not much to express from the _status quo_ side.

**Reply by adriangb** [reply]
Yep understandable.

My main thoughts for removing it is:
- It's better to be explicit. As we now know, this automatic wrapping can lead to surprising behavior with contextvars or detection of coroutines. This can confuse users, as evidenced by the multiple related issues, some of which have dozens of likes, and the fact that the resultant behavior is hard to explain even for advanced developers (e.g. Sebastian's comment in the upstream FastAPI issue, and frankly this stuff confuses me as well)
- It's 1 LOC / decorator for users to wrap their own functions.
- We would remove quite a bit of tricky logic from Starlette (what we have there now, but also what we save in the future by not merging #1444 or #1258)
